optimize/dataio.py

```diff
@@ -65,31 +67,51 @@ def __init__(self, filename, ntrack, swap_xz=True, seed=3, random_ntrack=False,
             for i_rand in list_rand:
                 fit_index.append(index[i_rand])
                 fit_tracks.append(all_tracks[i_rand])
-        
+      
+        if iterations is not None and iterations<max_nbatch or max_nbatch<0: 
+            max_nbatch = iterations
+
         if max_batch_len is not None:
             batches = []
             batch_here = []
             tot_length = 0
+            tot_data_length = 0
+            done_track_looping = False
             for track in fit_tracks:
                 for segment in track:
+                    if segment[self.track_fields.index("dx")] > max_batch_len:
+                        continue
                     tot_length+=segment[self.track_fields.index("dx")]
                     if tot_length < max_batch_len:
                         batch_here.append(segment)
                     else:
-                        batches.append(torch.stack(batch_here))
+                        if len(batch_here) > 0:
+                            batches.append(torch.stack(batch_here))
+                            tot_data_length += tot_length - segment[self.track_fields.index("dx")]
+                            print("batch length: ", tot_length - segment[self.track_fields.index("dx")])
                         batch_here = []
                         tot_length = 0
-
+                        if len(batches) >= max_nbatch and max_nbatch > 0: 
```

Comment:
Also style thing, but I would maybe make the case of no max `max_nbatch == None`. Which, now that I think of it, should maybe be done for `ntrack` too